sdks/python/apache_beam/io/iobase.py

```diff
@@ -1236,6 +1236,74 @@ def try_claim(self, position):
     raise NotImplementedError
 
 
+class WatermarkEstimator(object):
+  """A WatermarkEstimator which is used for estimating output_watermark based on
+  the timestamp of output records or manual modifications.
+
+  The base class provides common APIs that are called by the framework, which
+  are also accessible inside a DoFn.process() body. One derived watermark
+  estimator should implement all framework-called APIs, meanwhile, new added
+  APIs of the derived watermark estimator will also be exposed to a DoFn body.
+
+  Multi-threading safety is guarded by ThreadsafeWatermarkEstimator.
+  """
+  def get_estimator_state(self):
+    """Get current state of the WatermarkEstimator instance, which can be used
+    to recreate the WatermarkEstimator when processing residual.
+
+    This function is called by the system.
+    """
+    raise NotImplementedError(type(self))
+
+  def current_watermark(self):
+    """Get estimated output_watermark. This function is called by system."""
+    raise NotImplementedError(type(self))
+
+  def observe_timestamp(self, timestamp):
+    """Update tracking  watermark with latest output timestamp.
+
+    Args:
+      timestamp: the `timestamp.Timestamp` of current output element.
+
+    This function is called by system.
+    """
+    raise NotImplementedError(type(self))
+
+
+class ThreadsafeWatermarkEstimator(object):
```

Comment:
We should hide this as an implementation detail within the sdk harness code as we don't want users to have to worry about this.

Reply:
Yes. I'm going to move `ThreadsafeRestrictionTracker`, `RestrictionTrackerView` and `ThreadsafeWatermarkEstimator` to `common.py`